src/browser.ts

```diff
@@ -235,13 +307,95 @@ async function registerPlugins(
   return ctx
 }
 
+class PreInitMethodCallBuffer {
+  /**
+   *  Represents any and all the buffered method calls that occurred before initialization.
+   */
+  private _list: PreInitMethodCall[] = []
+
+  private _windowBuffer: PreInitMethodCall[] = []
+
+  public saveSnippetWindowBuffer() {
+    // store reference in order to be be to mutate 'method.called' field from callers.
+    this._windowBuffer = getSnippetWindowBuffer()
+  }
+
+  public get list(): PreInitMethodCall[] {
+    return [...this._list, ...this._windowBuffer]
+  }
+
+  push(...calls: PreInitMethodCall[]) {
+    this._list.push(...calls)
+  }
+
+  clear(): void {
+    this._list = []
+  }
+
+  filterByMethod<T extends PreInitMethodName>(
+    methodName: T
+  ): PreInitMethodCall<T>[] {
+    return this.list.filter(
+      (el): el is PreInitMethodCall<T> => el.method === methodName
+    )
+  }
+
+  /**
+   *  Call method and mark as "called"
+   */
+  async callMethod<T extends PreInitMethodName>(
+    analytics: Analytics,
+    methodCall: PreInitMethodCall<T>
+  ) {
+    if (methodCall.called) {
+      return undefined
+    }
+
+    methodCall.called = true
+
+    const { method, args, resolve } = methodCall
+    const result = await (analytics[method] as Function)(...args)
+    resolve(result)
+  }
+}
+
 export class AnalyticsBrowser {
+  /**
+   * Analytics class instance initialized by AnalyticsBrowser.load.
+   * Reinitializes everytime load is called.
+   */
+  static instance?: Analytics = undefined
+  /**
+   * Analytics context instance initialized by AnalyticsBrowser.load.
+   */
+  static ctx?: Context = undefined
+
+  static preInitBuffer = new PreInitMethodCallBuffer()
+
+  /**
+   * Clear the global state. Useful mainly for testing.
+   */
+  static _resetGlobalState() {
```

Comment:
Related to: https://github.com/segmentio/analytics-next/pull/391

I wish the codebase start to move away from these classes, especially if every method is a static method or variable.

Such practices make the codebase less tree-shakeable and treat the class as a bag of things (a file module) when you can achieve the same doing `import * as MyName from '...'` if you care about the aesthetic aspect of it.
Also, it is quite hard to follow, in the FP version you would have clear boundaries with data, state, and behavior.

I made the PR to showcase what it will look like and hopefully you would move there sooner than later. (Sentry@v7 is doing a lot of work on that front if you want validation, your products are really close when it comes to SDKs)

The bigger this class keeps getting, the harder it will be replace it or justify the breaking change, and the Sunk Cost Fallacy will be a thing.

My :2cents:

Reply:
Just chiming in with some context in regards to the Sentry side (https://github.com/getsentry/sentry-javascript)!

Static methods and attributes in classes are something we've stayed away from in our Browser SDK primarily because they just can't get minified. Something like `AnalyticsBrowser.trackSubmit()` minifies to `aa.trackSubmit()`, vs. for a function -> `function trackSubmit()` -> `function aa()`. This adds up over time (and suddenly makes function name length actually matter, never a good convo to have in code reviews). As an unrelated note, notice how this also applies for all objects in general, which makes destructuring and object keys all have a cost as well.

This is further exacerbated by the fact that the bundle increases even more when classes are transpiled down to es5 (though this is more an argument to just not use classes at all).

Then there's the general problem of bringing in unrelated logic whenever you import in the class to use static methods, but that is generally solved by keeping classes small + using dependency injection as much as possible (also just not using classes at all).

In general, we've had a lot of gains in moving private/static methods outside of a class (with shims for backwards compatibility until we can make the breaking change). See https://github.com/getsentry/sentry-javascript/pull/4281 as an example!

If you have any questions, please ping, happy to help out :)

Reply:
> Related to: #391
> 
> I wish the codebase start to move away from these classes, especially if every method is a static method or variable.
> 
> Such practices make the codebase less tree-shakeable and treat the class as a bag of things (a file module) when you can achieve the same doing `import * as MyName from '...'` if you care about the aesthetic aspect of it. Also, it is quite hard to follow, in the FP version you would have clear boundaries with data, state, and behavior.
> 
> I made the PR to showcase what it will look like and hopefully you would move there sooner than later. (Sentry@v7 is doing a lot of work on that front if you want validation, your products are really close when it comes to SDKs)
> 
> The bigger this class keeps getting, the harder it will be replace it or justify the breaking change, and the Sunk Cost Fallacy will be a thing.
> 
> My :2cents:

> > Somewhat related — I did some naive comparison testing in our playground using webpack-bundle-analyzer before / after (master vs this branch) and there was essentially no difference in the bundle size.
> 
> As I pointed out here: #391 is just a simple change already reducing the bundle size. Is it minimal? yes, most definitely. But for such minimal work that is exactly the same just changing the habits already pays off. As an application developer, I would appreciate every byte you save.
> 
> I could continue with the rabbit hole, as @AbhiPrasad pointed out, Sentry already proved some of these concepts with huge benefits.
> 
> I can try to do what I did on [getsentry/sentry-javascript#4381](https://github.com/getsentry/sentry-javascript/pull/4381) if you need more proof that it will eventually compound.
> 
> If I am not mistaken Sentry has seen up to 24% reduction and that is not even counting optimizations that you could continue doing, nor counting that it is not 100% of the work that could be done.
> 
> Anyway, it could a discussion for later, but as I pointed out "Sunk Cost" may make things harder to change, we are all humans after all.
> 
> ❤️

As usual, thanks @yordis for the input. I don't think the static methods precludes us from refactoring to a more "functional" and "tree-shaking" friendly interface (per your previous PR).

I can image us eventually refactoring the library to have an entry point that is something like:
```ts
export { track } from './track'
export { load } from './load'
// ... more methods


// for legacy users -- it looks like a class, but it's just a POJO
export const AnalyticsBrowser = {
   load, // or dynamic import?
   track,
// ... more methods 
}

```

 And thank you @AbhiPrasad, these is great feedback and I will definitely look @sentry's SDK moving forward.

Reply:
I’m available on [Sentry’s discord](https://discord.gg/Ww9hbqr) if you ever need anything.

Also apologizes if I came across like I was bikeshedding or causing friction on your PR - totally not the intention, just trying to share some learnings (and hopefully learn some stuff from you folks as well!)

Reply:
@AbhiPrasad no; you weren’t bikeshedding at all, this is great stuff to bring up!